### ISSUE TITLE
Intermittent AccessViolation exception in LsaIntrop.GetTicket causing application to crash

### ISSUE THREAD

**Issue by ElectricVampire** [SteveSyfuhs/Kerberos.NET#356]
**Describe the bug**
We are connecting oracle db from windows service running in .NET6 which internally usages Kerberos.NET. We saw that our service crashes intermittently. 
Unhandled exception at 0x00007FF7ECA3337E in Test.Service__PID__20180__Date__09_05_2023__Time_12_28_09PM__495__First Chance Access Violation.dmp: 0xC0000005: Access violation reading location 0x000001E0C19F0266.

After capturing the dump file I found that exception is comming on following code. **Here cred.ToArray() is throwing the access violation exception**

```
return WithFixedBuffer(response, p =>
                {
                    var pResponse = (KERB_RETRIEVE_TKT_RESPONSE*)p;

                    var cred = new Span<byte>(pResponse->Ticket.EncodedTicket, pResponse->Ticket.EncodedTicketSize);

                    return KrbCred.DecodeApplication(cred.ToArray());
                });
```

**Dump analysis**

I switched to thread id where exception occurred. Looked at stack trace and identified that issue can come if LsaBufferSafeHandle is pointing to invalid pointer. When dumped the object handle is 0. Which will make source as null in ToArray call causing the exception.
I don't know why its coming null sometime. Please help us analyzing the issue.

```
> setthread 31

> clrstack
OS Thread Id: 0x42b4 (31)
        Child SP               IP Call Site
000000BFEBDFE210 00007FF7ECA3337E System.Buffer.Memmove(Byte ByRef, Byte ByRef, UIntPtr) [/_/src/libraries/System.Private.CoreLib/src/System/Buffer.cs @ 269]
000000BFEBDFE270 00007FF7EF7693F7 System.Span`1[[System.Byte, System.Private.CoreLib]].ToArray() [/_/src/libraries/System.Private.CoreLib/src/System/Span.cs @ 435]
000000BFEBDFE2B0 00007FF7EF80F01A Kerberos.NET.Win32.LsaInterop+<>c.<GetTicket>b__19_1(IntPtr) [D:\a\1\s\Kerberos.NET\Win32\LsaInterop.cs @ 360]
000000BFEBDFE310 00007FF7EF80EED3 Kerberos.NET.Win32.LsaInterop.WithFixedBuffer[[System.__Canon, System.Private.CoreLib]](Kerberos.NET.Win32.LsaBufferSafeHandle, System.Func`2<IntPtr,System.__Canon>) [D:\a\1\s\Kerberos.NET\Win32\LsaInterop.cs @ 369]
000000BFEBDFE370 00007FF7EF80C48C Kerberos.NET.Win32.LsaInterop.WithFixedBuffer[[System.__Canon, System.Private.CoreLib]](Int32, System.Func`3<IntPtr,Int32,System.__Canon>) [D:\a\1\s\Kerberos.NET\Win32\LsaInterop.cs @ 381]
000000BFEBDFE410 00007FF7EF74E410 Kerberos.NET.Client.LsaCredentialCache.TryWithKerberosCatch[[System.__Canon, System.Private.CoreLib]](System.Func`1<System.__Canon>, System.Func`2<Kerberos.NET.KerberosProtocolException,System.__Canon>) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 178]
000000BFEBDFE460 00007FF7EF80C20B Kerberos.NET.Client.LsaCredentialCache.GetTicketOrThrow(System.String) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 167]
000000BFEBDFE4A0 00007FF7EF74E410 Kerberos.NET.Client.LsaCredentialCache.TryWithKerberosCatch[[System.__Canon, System.Private.CoreLib]](System.Func`1<System.__Canon>, System.Func`2<Kerberos.NET.KerberosProtocolException,System.__Canon>) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 178]
000000BFEBDFE4F0 00007FF7EED24CDF Kerberos.NET.Client.LsaCredentialCache.TryGetTicketOrThrow(System.String) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 144]
000000BFEBDFE580 00007FF7EF74E410 Kerberos.NET.Client.LsaCredentialCache.TryWithKerberosCatch[[System.__Canon, System.Private.CoreLib]](System.Func`1<System.__Canon>, System.Func`2<Kerberos.NET.KerberosProtocolException,System.__Canon>) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 178]
000000BFEBDFE5D0 00007FF7EED241E4 Kerberos.NET.Client.LsaCredentialCache.GetCacheItem(System.String, System.String) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 94]
000000BFEBDFE630 00007FF7EF80EB07 Kerberos.NET.Client.LsaCredentialCache.GetCacheItem[[Kerberos.NET.Client.KerberosClientCacheEntry, Kerberos.NET]](System.String, System.String) [D:\a\1\s\Kerberos.NET\Cache\LsaCredentialCache.cs @ 201]
000000BFEBDFE6F0 00007FF7EED20032 Kerberos.NET.Client.KerberosClient+<GetServiceTicket>d__79.MoveNext() [D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs @ 575]
000000BFEBDFEA60 00007FF7EF80E5E8 System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Kerberos.NET.Client.KerberosClient+<GetServiceTicket>d__79, Kerberos.NET]](<GetServiceTicket>d__79 ByRef) [/_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs @ 38]
000000BFEBDFEAC0 00007FF7EF80E541 Kerberos.NET.Client.KerberosClient.GetServiceTicket(Kerberos.NET.RequestServiceTicket, System.Threading.CancellationToken)
000000BFEBDFEC50 00007FF7EF80CE03 OracleInternal.Network.AnoKerberos.Authenticate(System.String, System.String, System.String, System.String, Byte[] ByRef)
000000BFEBDFEDF0 00007FF7EECFE9ED OracleInternal.Network.Ano.KerberosHandshake(OracleInternal.Network.AuthenticationService, System.String, System.String)
000000BFEBDFEE70 00007FF7EE772B0B OracleInternal.Network.OracleCommunication.SendConnectPacketAndProcessResponse(OracleInternal.Network.AddressResolution)
000000BFEBDFEEF0 00007FF7EE76BBC1 OracleInternal.Network.OracleCommunication.ConnectViaCO(OracleInternal.Network.ConnectionOption, OracleInternal.Network.AddressResolution)
000000BFEBDFEF60 00007FF7EE772D20 OracleInternal.Network.OracleCommunication.SendConnectPacketAndProcessResponse(OracleInternal.Network.AddressResolution)
000000BFEBDFEFE0 00007FF7EE76BBC1 OracleInternal.Network.OracleCommunication.ConnectViaCO(OracleInternal.Network.ConnectionOption, OracleInternal.Network.AddressResolution)
000000BFEBDFF050 00007FF7EE75C774 OracleInternal.Network.OracleCommunication.DoConnect(System.String)
000000BFEBDFF0F0 00007FF7EF8071B7 OracleInternal.ServiceObjects.OracleConnectionImpl.Connect(OracleInternal.ConnectionPool.ConnectionString, Boolean, Oracle.ManagedDataAccess.Client.OracleConnection, System.String)
000000BFEBDFF2E0 00007FF7EF80AB1F OracleInternal.ConnectionPool.PoolManager`3[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].CreateNewPRThreadFunc(System.Object) 000000BFEBDFF370 00007FF7EF80B238 System.Threading.QueueUserWorkItemCallback.Execute() [/_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs @ 923]
000000BFEBDFF3A0 00007FF7ED643961 System.Threading.ThreadPoolWorkQueue.Dispatch() [/_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs @ 724]
000000BFEBDFF430 00007FF7ED641C20 System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart() [/_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs @ 63]
000000BFEBDFF780 00007ff84c5caed3 [DebuggerU2MCatchHandlerFrame: 000000bfebdff780]
> dso
OS Thread Id: 0x42b4 (31)
RSP/REG          Object           Name
000000BFEBDFE230 000001e0a1db4558 Kerberos.NET.Client.LsaCredentialCache+<>c__DisplayClass16_0
000000BFEBDFE248 000001e0a1dd1118 System.Byte[]
000000BFEBDFE280 000001e0a1db4558 Kerberos.NET.Client.LsaCredentialCache+<>c__DisplayClass16_0
000000BFEBDFE290 000001e0a1db4558 Kerberos.NET.Client.LsaCredentialCache+<>c__DisplayClass16_0
000000BFEBDFE298 000001e09da48578 System.Func`2[[System.IntPtr, System.Private.CoreLib],[Kerberos.NET.Entities.KrbCred, Kerberos.NET]]
000000BFEBDFE2A0 000001e0a1db45c0 System.Func`3[[System.IntPtr, System.Private.CoreLib],[System.Int32, System.Private.CoreLib],[Kerberos.NET.Entities.KrbCred, Kerberos.NET]]
000000BFEBDFE2B0 000001e0a1db4600 Kerberos.NET.Win32.LsaBufferSafeHandle
000000BFEBDFE2D0 000001e0a1db00a0 Kerberos.NET.Win32.LsaSafeHandle
000000BFEBDFE318 000001e0a1db4600 Kerberos.NET.Win32.LsaBufferSafeHandle
000000BFEBDFE388 000001e09d379dd8 System.Globalization.CompareInfo
000000BFEBDFE3A0 000001e0a1db64e0 System.Security.Cryptography.CryptoMemoryOwner`1[[System.Byte, System.Private.CoreLib]]
000000BFEBDFE3D0 000001e0a1db63c8 System.Byte[]
000000BFEBDFE3E0 000001e0a1db63c8 System.Byte[]
000000BFEBDFE3F0 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFE448 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFE480 000001e0a1db43a0 Kerberos.NET.Client.LsaCredentialCache+<>c__DisplayClass14_0
000000BFEBDFE490 000001e0a1db44d8 System.Func`1[[Kerberos.NET.Entities.KrbCred, Kerberos.NET]]
000000BFEBDFE500 000001e0a1db4518 System.Func`2[[Kerberos.NET.KerberosProtocolException, Kerberos.NET],[Kerberos.NET.Entities.KrbCred, Kerberos.NET]]
000000BFEBDFE528 000001e0a1db43c8 System.Collections.Generic.List`1[[System.String, System.Private.CoreLib]]
000000BFEBDFE530 000001e0a1db42d0 System.String    oracle/db.service.net
000000BFEBDFE540 000001e0a1d9f6c8 System.String    domain.net
000000BFEBDFE558 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFE5E0 000001e0a1db4360 System.Func`2[[Kerberos.NET.KerberosProtocolException, Kerberos.NET],[Kerberos.NET.Entities.KrbCred, Kerberos.NET]]
000000BFEBDFE618 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFE620 000001e0a1db42d0 System.String    oracle/db.service.net
000000BFEBDFE6D0 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFE6E0 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFE6F0 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFE730 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFEA40 000001e0a1d88818 System.String    db.service.net
000000BFEBDFEA98 000001e09f4aa2e8 System.Threading.ExecutionContext
000000BFEBDFEAA0 000001e0a0b0d060 System.Threading.Thread
000000BFEBDFEAE0 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFEAE8 000001e0a1db41c0 Kerberos.NET.Entities.KrbPrincipalName
000000BFEBDFEAF0 000001e0a1db4208 System.String    krbtgt/domain.net
000000BFEBDFEAF8 000001e09da44008 Microsoft.Extensions.Logging.NullScope
000000BFEBDFEB20 000001e0a1db40c8 System.String    oracle/db.service.net
000000BFEBDFEB58 000001e0a1da1550 Kerberos.NET.Configuration.Krb5Config
000000BFEBDFEC30 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFEC50 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFEC58 000001e0a1daff68 Kerberos.NET.Client.LsaCredentialCache
000000BFEBDFEC60 000001e09d738e40 System.String    MSLSA:
000000BFEBDFEC98 000001e0a1daeff0 Kerberos.NET.Client.KerberosClient
000000BFEBDFECA0 000001e0a1db40c8 System.String    oracle/db.service.net
000000BFEBDFED30 000001e0a1db40c8 System.String    oracle/db.service.net
000000BFEBDFEDA8 000001e09d738e40 System.String    MSLSA:
000000BFEBDFEDB0 000001e09f517ed8 OracleInternal.Network.Ano
000000BFEBDFEDB8 000001e09d738d88 System.String    Test\oracle\krb5.conf
000000BFEBDFEDC8 000001e0a1d88818 System.String    db.service.net
000000BFEBDFEDD0 000001e09f517f98 OracleInternal.Network.AuthenticationService
000000BFEBDFEDD8 000001e0a1d887f0 System.String    oracle
000000BFEBDFEDF0 000001e09d738d88 System.String    Test\oracle\krb5.conf
000000BFEBDFEDF8 000001e09f517ed8 OracleInternal.Network.Ano
000000BFEBDFEE00 000001e09f517f98 OracleInternal.Network.AuthenticationService
000000BFEBDFEE08 000001e09f518350 OracleInternal.Network.ConnectPacket
000000BFEBDFEE30 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFEE38 000001e09e90cf00 OracleInternal.Network.AddressResolution
000000BFEBDFEE50 000001e09f518350 OracleInternal.Network.ConnectPacket
000000BFEBDFEEB0 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFEF38 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFEF40 000001e09e90cf00 OracleInternal.Network.AddressResolution
000000BFEBDFEF60 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFEF70 000001e09e90cf00 OracleInternal.Network.AddressResolution
000000BFEBDFEFA0 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFEFF8 000001e09d361240 System.String
000000BFEBDFF020 000001e09d361240 System.String
000000BFEBDFF028 000001e09dd7b950 OracleInternal.Network.ConnectionOption
000000BFEBDFF030 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFF038 000001e09f461a90 Oracle.ManagedDataAccess.Client.OracleConnection
000000BFEBDFF050 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFF060 000001e09e90cf00 OracleInternal.Network.AddressResolution
000000BFEBDFF0B0 000001e09f6fea08 OracleInternal.Network.AddressResolution+<GetEnumerator>d__148
000000BFEBDFF0B8 000001e09e90cf00 OracleInternal.Network.AddressResolution
000000BFEBDFF0C8 000001e09d7392d8 OracleInternal.ConnectionPool.ConnectionString
000000BFEBDFF0D0 000001e09f6fdf90 OracleInternal.ServiceObjects.OracleConnectionImpl
000000BFEBDFF0F0 000001e09e90ccf0 OracleInternal.Network.OracleCommunication
000000BFEBDFF158 000001e09d3a65c0 System.String    /
000000BFEBDFF1E8 000001e09d69c180 System.Collections.Hashtable
000000BFEBDFF208 000001e09e90c300 OracleInternal.Common.SqlNetOraConfig
000000BFEBDFF298 000001e09f6fdf90 OracleInternal.ServiceObjects.OracleConnectionImpl
000000BFEBDFF2B0 000001e09f6fe948 System.Threading.QueueUserWorkItemCallback
000000BFEBDFF2C0 000001e09f461a90 Oracle.ManagedDataAccess.Client.OracleConnection
000000BFEBDFF2E0 000001e09f6fdf90 OracleInternal.ServiceObjects.OracleConnectionImpl
000000BFEBDFF318 000001e09f6fe8d8 System.Threading.ManualResetEventSlim
000000BFEBDFF320 000001e09f6fe8a0 System.Threading.ManualResetEvent
000000BFEBDFF328 000001e09f6fdf90 OracleInternal.ServiceObjects.OracleConnectionImpl
000000BFEBDFF338 000001e09d99e050 OracleInternal.ConnectionPool.OraclePoolManager
000000BFEBDFF340 000001e0a0b0d3c0 System.Threading.ThreadInt64PersistentCounter+ThreadLocalNode
000000BFEBDFF348 000001e09d47ca90 System.Threading.ThreadPoolWorkQueue
000000BFEBDFF350 000001e0a0b0d210 System.Threading.ThreadPoolWorkQueueThreadLocals
000000BFEBDFF358 000001e0a0b0d060 System.Threading.Thread
000000BFEBDFF370 000001e09d99e050 OracleInternal.ConnectionPool.OraclePoolManager
000000BFEBDFF388 000001e09d47ca90 System.Threading.ThreadPoolWorkQueue
000000BFEBDFF390 000001e09f6fe948 System.Threading.QueueUserWorkItemCallback
000000BFEBDFF3D0 000001e09d47ca90 System.Threading.ThreadPoolWorkQueue
000000BFEBDFF3F8 000001e09d47d2e0 System.Threading.PortableThreadPool
000000BFEBDFF4B0 000001e09d47d708 System.Threading.LowLevelLifoSemaphore
000000BFEBDFF4B8 000001e09d47d550 System.Threading.LowLevelLock
000000BFEBDFF4C0 000001e09d47d2e0 System.Threading.PortableThreadPool


> dumpobj 000001e0a1db4600
Name:        Kerberos.NET.Win32.LsaBufferSafeHandle
MethodTable: 00007ff7eef7c910
EEClass:     00007ff7eef86360
Tracked Type: false
Size:        40(0x28) bytes
File:        Test.Service\Kerberos.NET.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ff7ecac4298  40011b9        8        System.IntPtr  1 instance 0000000000000000 handle
00007ff7ecab94c0  40011ba       10         System.Int32  1 instance                3 _state
00007ff7ec9fbf60  40011bb       14       System.Boolean  1 instance                1 _ownsHandle
00007ff7ec9fbf60  40011bc       15       System.Boolean  1 instance                1 _fullyInitialized
00007ff7ecab94c0  400010c       18         System.Int32  1 instance             2336 <BufferLength>k__BackingField
```

**Comment by SteveSyfuhs** [comment]
Which package version is this running? I recently fixed a memory leak here: https://github.com/dotnet/Kerberos.NET/commit/cdaf240f8cbd206e2102c990ea22f186002d7441

It would periodically pop up as weird pointer errors under load.

**Comment by ElectricVampire** [comment]
Version of Kerberos.NET.dll is 4.6.13.2847. Which version has this fix?

**Comment by SteveSyfuhs** [comment]
Anything after https://www.nuget.org/packages/Kerberos.NET/4.6.26 should have the fix.

It would probably be useful to add a null handle check in that path to make the error more explicit, though beyond that I don't specifically know why it would be null. That would suggest `LsaCallAuthenticationPackage` is not setting the handle correctly, but if that were the case I'd also expect it to return an error code, which `LsaThrowIfError` should be dealing with.

**Comment by ElectricVampire** [comment]
I'll use the latest version and see if we get the issue.
Just want to point out that buffer len is set which means call was success, Is there any possibility that buffer was freed from some other place? If you need any object value from any of frames in stack trace to get more information around the issue, I can share it.

**Comment by SteveSyfuhs** [comment]
No, the handle should be safe from anyone else touching it. It's local to the parent function and passed directly to the caller. It's never shared with anyone.

**Comment by ElectricVampire** [comment]
The issue for which [change ](https://github.com/dotnet/Kerberos.NET/commit/cdaf240f8cbd206e2102c990ea22f186002d7441) was made was giving AccessViolation exception? Just trying to see if this change will solve our issue.

I'll update the nuget and see if it solves the problem. If not what changes/logging we can do more in Kerberos.NET to get more information to know the cause, so when it happens again we have more details around it,

**Comment by SteveSyfuhs** [comment]
Specifically putting it in a using block. Not having it in the using block was leaking the handle data, possibly until the finalizer kicked in, but in between the finalizer executing and the `fixed` block exiting, it's plausible something got moved and paved over a separate handle. I don't know if they're related issues. I just noticed fixing the leak also fixed some random native exceptions when run under load.
